pkg/db/db.go

```diff
@@ -0,0 +1,117 @@
+package db
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+	"go.uber.org/zap"
+)
+
+func GetConnectionString(host, user, password, name string, port int) string {
+	return fmt.Sprintf(
+		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
+		host,
+		port,
+		user,
+		password,
+		name,
+	)
+}
+
+type Database struct {
+	tables struct {
+		providers string
+	}
+	pool *pgxpool.Pool
+	// conn is used to listen for notifications and should be closed
+	conn *pgxpool.Conn
+}
+
+func NewClient(ctx context.Context, connString string, tables map[string]string) (*Database, error) {
+	if _, ok := tables["providers"]; !ok {
+		return nil, fmt.Errorf("incomplete table name mapping, missing a table name for \"providers\" table")
+	}
+
+	pool, err := pgxpool.New(ctx, connString)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new instace of pool: %v", err)
+	}
+
+	if err := pool.Ping(ctx); err != nil {
+		return nil, fmt.Errorf("failed to ping database: %v", err)
+	}
+
+	db := Database{}
+	db.pool = pool
+	db.tables.providers = tables["providers"]
+	return &db, nil
+}
+
+func (pg *Database) Close(ctx context.Context) {
+	if pg.conn != nil {
+		pg.conn.Conn().Close(ctx)
+	}
+	pg.pool.Close()
+}
+
+func (pg *Database) OnProvider(ctx context.Context, logger *zap.Logger, notifyExisting bool, fn func(providerName string)) error {
+	var err error
+	if pg.conn, err = pg.pool.Acquire(ctx); err != nil {
+		return fmt.Errorf("failed to acquire connection: %v", err)
+	}
+
+	query := fmt.Sprintf("LISTEN %s;", pg.tables.providers)
+	if _, err = pg.conn.Exec(ctx, query); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			notification, err := pg.conn.Conn().WaitForNotification(ctx)
+			if err != nil {
+				// If the connection is closed, release the connection and exit the goroutine
+				if pg.conn.Conn().IsClosed() {
+					pg.conn.Release()
+					logger.Error("Connection closed by server:", zap.Error(err))
+					return
+				}
+				logger.Error("failed to wait for notification", zap.Error(err))
+				continue
+			}
+			fn(notification.Payload)
+		}
+	}()
+
+	if notifyExisting {
```

Comment:
- I believe this'll be the last comment from me, I don't foresee any scenario where this'll be false, do you see any?

Reply:
This might get change in the future for now, as @nb-ohad thinks we might have a scenario like that going forward we might need to keep it and remove it later if not at all required.

Reply:
@leelavg you are raising a good point. Right now nothing prevents us from calling OnProvider more than once. my assumption was that when we lose the connection (for one or another) we will want to error handle and reestablish the notification channel without returning the reconcile for all records.

Reply:
Ack, I was thinking more like to reconcile all records with duration let's say every 10 min and accepting duration seems more suitable.

Nevertheless, this isn't a blocker for now.